test/specs/services/faye.spec.js

```diff
@@ -32,26 +31,61 @@ const sandbox = sinon.sandbox.create();
 
 describe('Faye service', () => {
     let mockedStore;
+    let getMessagesStub;
+    let disconnectFayeStub;
+    let handleConversationUpdatedStub;
+    let showSettingsStub;
+    let hideChannelPageSpy;
+    let hideConnectNotificationSpy;
+    let addMessageSpy;
+    let incrementUnreadCountSpy;
+    let resetUnreadCountSpy;
+    let setUserSpy;
+    let setFayeConversationSubscriptionSpy;
+    let setFayeUserSubscriptionSpy;
 
     beforeEach(() => {
         mockedStore = createMockedStore(sandbox, getProps());
 
         sandbox.stub(Client.prototype, 'addExtension');
-        sandbox.stub(conversationService, 'getMessages');
-        sandbox.stub(conversationService, 'disconnectFaye');
-        sandbox.stub(conversationService, 'handleConversationUpdated');
-        sandbox.stub(appService, 'showSettings');
-        sandbox.stub(appService, 'hideChannelPage');
-        sandbox.stub(appService, 'hideConnectNotification');
-
-        sandbox.stub(conversationActions, 'addMessage');
-        sandbox.stub(conversationActions, 'incrementUnreadCount');
-        sandbox.stub(conversationActions, 'resetUnreadCount');
-        sandbox.stub(userActions, 'setUser');
-        sandbox.stub(fayeActions, 'setFayeConversationSubscription');
-        sandbox.stub(fayeActions, 'setFayeUserSubscription');
-
-        sandbox.stub(utilsDevice, 'getDeviceId').returns(123);
+
+        getMessagesStub = sandbox.stub().returns(() => Promise.resolve());
+        FayeRewire('getMessages', getMessagesStub);
```

Comment:
> but then hot-swap the reference inside the faye module with a stubbed version?
Yeah that's it and from what I understand, you can't to rewire 2 levels deeper. It only work for that module directly (maybve I'm wrong, but it looked like that when I tested it).

> traditional mock where you replace the implementation of a function, what you're really replacing is the faye module's internal reference to that function. Is that right?

if you look at what jest does (https://facebook.github.io/jest/), it stubs by default any dependencies of a module and you have to tell it which ones to let go through unstubbed. imo it makes a much more isolated test case without potential side effects if you were to mock 3 levels down.